### PR TITLE
Hotfix duplicate moves

### PR DESCRIPTION
**Pull request by ngw**
The old fix dealing with duplicate moves has been ported to deal with the new `nomis_event_ids` field

### CODE REVIEW

File: app/services/moves/sweeper.rb [hunk 1]
```diff
@@ -36,6 +36,8 @@ def update_nomis_event_ids!
         date: date,
         from_location_id: locations.map(&:id)
       )
+
+      update_nomis_event_ids_when_duplicate(scope)
```
Comment (spikeheap) [inline app/services/moves/sweeper.rb]: Are we using the same logic to detect potential duplicates as the unique index? Should this (L35-38 above) also have to_location and time_due?

File: spec/services/moves/sweeper_spec.rb [hunk 2]
```diff
@@ -54,117 +56,105 @@
...
-        it 'removes the outdated nomis_event_id' do
-          sweeper.call
-          expect(multiple_nomis_ids_move.reload.nomis_event_ids).to eq([attributes[:nomis_event_id]])
-        end
+      it 'cancels the outdated move' do
+        sweeper.call
+        expect(outdated_move.reload.status).to eq('cancelled')
       end
     end
   end
```
Comment (spikeheap) [inline spec/services/moves/sweeper_spec.rb]: Do we need an additional test case where there are no NOMIS ids? I may have missed this one...

File: app/services/moves/sweeper.rb [hunk 3]
```diff
@@ -48,6 +50,15 @@ def update_nomis_event_ids_containing_currents(scope)
       end
     end
 
+    def update_nomis_event_ids_when_duplicate(scope)
+      items.map do |item|
+        move = scope.find_by(
+          person: Person.where(nomis_prison_number: item[:person_nomis_prison_number])
+        )
+        move.update(nomis_event_ids: move.nomis_event_ids << item[:nomis_event_id])
```
[the remark this answers is not in the document]
Comment (ngw) [reply app/services/moves/sweeper.rb]: Yes, if you look into the model it will call uniq before validation

File: app/services/moves/sweeper.rb (hunk 1 above)
Comment (ngw) [reply app/services/moves/sweeper.rb]: The whole sweeper has been built around the concept that from_location + date + person = same move, this is the logic that was there before and the spec. Eventually using a timestamp could be a very bad idea, it normally gives a lot of problems.
If you guys think the to_location is relevant so be it

File: spec/services/moves/sweeper_spec.rb (hunk 2 above)
Comment (ngw) [reply spec/services/moves/sweeper_spec.rb]: I thought it was implicit that that case was covered, because the one you're commenting works (id gets removed from the array, then state changes to cancelled because the array is empty). Makes sense?

Comment (spikeheap) [reply spec/services/moves/sweeper_spec.rb]: Sorry, I meant for moves that never belonged to NOMIS (i.e. the data we lost before)

File: app/services/moves/sweeper.rb (hunk 1 above)
Comment (spikeheap) [reply app/services/moves/sweeper.rb]: Hi @ngw, from the spec doc:

> A Person could have multiple Moves on the same date, given they were for a different purpose (eg. a Move to court in the morning, and a Hospital Move in the evening).

It's important the uniqueness validation matches this logic, so I'll add a note on this PR next to the schema

File: db/migrate/20191106121438_unique_index_on_moves.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class UniqueIndexOnMoves < ActiveRecord::Migration[5.2]
+  def change
+    add_index :moves, [:from_location_id, :to_location_id, :person_id, :date, :time_due], name: 'index_on_move_uniqueness', unique: true
```
Comment (spikeheap) [inline db/migrate/20191106121438_unique_index_on_moves.rb]: Given https://github.com/ministryofjustice/hmpps-book-secure-move-api/pull/156#discussion_r343683943, this index should be `:from_location_id, :to_location_id, :person_id, :date`.

File: spec/services/moves/sweeper_spec.rb (hunk 2 above)
Comment (ngw) [reply spec/services/moves/sweeper_spec.rb]: There's no clear way to identify that, they all have empty nomis_event_ids. We have same move from nomis and outside of it now?

File: lib/tasks/data_maintenance.rake [hunk 5]
```diff
@@ -1,6 +1,17 @@
...
+  task remove_duplicate_moves: :environment do
+    duplicates = Move.select(:nomis_event_ids).group(:nomis_event_ids).having('COUNT(*) > 1').size
+    duplicates.each_pair do |k, _|
+      next if k.nil?
+
+      moves = Move.order(:created_at).where(nomis_event_ids: k).offset(1)
+      moves.destroy_all
+    end
+  end
+
```
Comment (spikeheap) [inline lib/tasks/data_maintenance.rake]: Hi @ngw, could we add a test for this please? I'm probably reading this wrong, but it looks like `duplicates` is set to be a number (the result of `.size`).

Could we also add some comments, and name variables more explicitly. It's not clear to me what `k` represents, and why it's pulled out with `each_pair`.

File: spec/services/moves/sweeper_spec.rb (hunk 2 above)
Comment (jason-sims) [reply spec/services/moves/sweeper_spec.rb]: There is a really clear way to identify locations that should be synced with NOMIS. If the NOMIS client is hitting the **movements** api requesting moves from.. lets say HMP Pentonville (PVI) then this site and all the moves originating from it, should be synced. The movements api is key as this tells us users are not entering moves in our service for that site.(they also have the pecs_prison role & caseload - our workflow doesn't allow us create moves in our service) What sites a user seeing is due to HMPPS Authentication (If it is an Auth user (outside of NOMIS) its a police station, If its non Auth (inside NOMIS - we have to do a secondary lookup to get caseload) then its a Prison). Not all requests from the frontend invoke the NOMIS lookup - all that logic is in the codebase.

Comment (spikeheap) [reply spec/services/moves/sweeper_spec.rb]: :+1: @ngw this might be what you're after: https://github.com/ministryofjustice/hmpps-book-secure-move-api/blob/master/app/models/location.rb#L21-L23

File: lib/tasks/data_maintenance.rake (hunk 5 above)
Comment (spikeheap) [reply lib/tasks/data_maintenance.rake]: Morning :Wave:. I've dug into this code a bit more and I think the behaviour is correct. I'd recommend refactoring it to make it more readable, for example:

```
    duplicate_event_ids = Move
                            .group(:nomis_event_ids)
                            .having('COUNT(*) > 1')
                            .pluck(:nomis_event_ids)
                            .compact
    
    duplicate_event_ids.each  do |event_ids|
      Move
        .where(nomis_event_ids: event_ids)
        .order(:created_at)
        .offset(1)
        .destroy_all
    end
```

I also think this needs a test, as it destroys data.

File: spec/services/moves/sweeper_spec.rb (hunk 2 above)
Comment (ngw) [reply spec/services/moves/sweeper_spec.rb]: Ok, but then the logic we currently have is 100% correct and covering all the cases, right? The sweeper does not act on "all the moves", but limited to moves that are passed as arguments. Once the cancellation of the moves is scoped on the locations provided (which are provided directly by nomis) we should be perfectly safe.

Comment (spikeheap) [reply spec/services/moves/sweeper_spec.rb]: Hi @ngw, I think we should just merge it at this point, but your comment above makes me feel quite uneasy. We should be coding defensively, particularly on this sort of thing where important data is being modified. The assumption that it's "100% correct" relies on the logic in another class not changing, which means we have a dependency on the thing that invokes this... it's easy to foresee this being called somewhere else in future, and without that defensive check we _could_ alter data in a way we might not expect. 

It's good to be kind to our future selves and protect ourselves from that sort of error.

As an example, you could write a test case here that invokes the sweeper for a location which is not a prison and has no `nomis_event_ids`. Even if we don't add that test before merging this (and I believe @jason-sims is keen to manually test this in staging today), I'd like to see this test in a subsequent PR.
